File: Stack and Queues/Min Stack.cpp
class MinStack {
private:
    vector<pair<int, int>> minStack;
public:
    MinStack() {}
    void push(int x) {
        if(minStack.empty()) minStack.emplace_back(x,x);
        else minStack.emplace_back(x, min(minStack.back().second, x));
    }
    void pop() {
        minStack.pop_back();
    }
    int top() {
        return minStack.back().first;
    }
    int getMin() {
        return minStack.back().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Stack and Queues/Min Stack.rs
struct MinStack {
    stack: Vec<(i32, i32)>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {

    fn new() -> Self {
        Self {
            stack: Vec::new(),
        }
    }
    
    fn push(&mut self, val: i32) {
        self.stack.push(match self.stack.last() {
            Some(&(_, min)) => {
                (val, min.min(val))
            },
            None => (val, val),
        })
    }
    
    fn pop(&mut self) {
        self.stack.pop();
    }
    
    fn top(&self) -> i32 {
        self.stack.last().unwrap().0
    }
    
    fn get_min(&self) -> i32 {
        self.stack.last().unwrap().1
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(val);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.get_min();
 */